File: dashboard/src/lib/routing.ts
// Route alternatives from OSRM's free public demo server, ranked by how many CURRENT flooded
// reports lie along each route. OSRM knows roads, not floods: a road with no report simply
// looks clear, so this ranks by the reports we have, it does not certify a route as safe.
// Demo-only dependency: the public server is best-effort. Self-host OSRM (or swap the URL)
// if this ever matters beyond a demo.
import { Item } from './items';

const OSRM = 'https://router.project-osrm.org/route/v1/driving';
/** A report within this many metres of the route line counts as "along" it. */
const NEAR_M = 75;

export interface LatLon {
  lat: number;
  lon: number;
}

export interface FloodPoint extends LatLon {
  sev: 'S2' | 'S3' | 'SX';
}

export interface RouteOption {
  coords: [number, number][]; // [lon, lat]
  distanceM: number;
  durationS: number;
  s3: number;
  s2: number;
  sx: number;
  safest: boolean;
}

/** Current flooded reports worth avoiding: S3 (impassable), S2 (impassable for cars), SX (conflicting). */
export function floodPoints(items: Item[]): FloodPoint[] {
  const out: FloodPoint[] = [];
  for (const i of items) {
    if (i.kind !== 'report' || i.stale) continue;
    const sev = i.summary.severity;
    if (sev === 'S3' || sev === 'S2' || sev === 'SX') out.push({ lat: i.lat, lon: i.lon, sev });
  }
  return out;
}

/** Metres from point p to segment a-b, on a local flat approximation (fine at city scale). */
function distToSegmentM(p: LatLon, a: [number, number], b: [number, number]): number {
  const kx = 111_320 * Math.cos((p.lat * Math.PI) / 180);
  const ky = 110_540;
  const ax = (a[0] - p.lon) * kx, ay = (a[1] - p.lat) * ky;
  const bx = (b[0] - p.lon) * kx, by = (b[1] - p.lat) * ky;
  const dx = bx - ax, dy = by - ay;
  const len2 = dx * dx + dy * dy;
  const t = len2 === 0 ? 0 : Math.max(0, Math.min(1, -(ax * dx + ay * dy) / len2));
  return Math.hypot(ax + t * dx, ay + t * dy);
}

function nearRoute(p: LatLon, coords: [number, number][]): boolean {
  for (let i = 1; i < coords.length; i++) {
    if (distToSegmentM(p, coords[i - 1], coords[i]) <= NEAR_M) return true;
  }
  return false;
}

const score = (r: RouteOption) => r.s3 * 3 + r.s2 * 2 + r.sx;

export async function fetchRoutes(from: LatLon, to: LatLon, floods: FloodPoint[]): Promise<RouteOption[]> {
  const url = `${OSRM}/${from.lon},${from.lat};${to.lon},${to.lat}?alternatives=3&overview=full&geometries=geojson&steps=false`;
  let res: Response;
  try {
    res = await fetch(url);
  } catch {
    throw new Error("Couldn't reach the routing service. Check the internet connection.");
  }
  if (!res.ok) throw new Error(`Routing service error (${res.status})`);
  const body = (await res.json()) as {
    code: string;
    routes?: { distance: number; duration: number; geometry: { coordinates: [number, number][] } }[];
  };
  if (body.code !== 'Ok' || !body.routes?.length) throw new Error('No road route was found between here and that location.');

  // Reports at the destination are the reason for going there, not an obstacle on the way.
  const obstacles = floods.filter((f) => Math.hypot((f.lat - to.lat) * 110_540, (f.lon - to.lon) * 111_320 * Math.cos((to.lat * Math.PI) / 180)) > NEAR_M);

  const options: RouteOption[] = body.routes.map((r) => {
    const coords = r.geometry.coordinates;
    const hit = obstacles.filter((f) => nearRoute(f, coords));
    return {
      coords,
      distanceM: r.distance,
      durationS: r.duration,
      s3: hit.filter((f) => f.sev === 'S3').length,
      s2: hit.filter((f) => f.sev === 'S2').length,
      sx: hit.filter((f) => f.sev === 'SX').length,
      safest: false,
    };
  });
  options.sort((a, b) => score(a) - score(b) || a.durationS - b.durationS);
  options[0].safest = true;
  return options;
}

export function currentPosition(): Promise<LatLon> {
  return new Promise((resolve, reject) => {
    if (!('geolocation' in navigator)) return reject(new Error("This browser can't share a location."));
    navigator.geolocation.getCurrentPosition(
      (p) => resolve({ lat: p.coords.latitude, lon: p.coords.longitude }),
      (e) =>
        reject(
          new Error(
            e.code === e.PERMISSION_DENIED
              ? 'Location is blocked for this page. Allow it in the browser to get routes from where you are.'
              : "Couldn't get your location.",
          ),
        ),
      { enableHighAccuracy: true, timeout: 10_000 },
    );
  });
}

/** Where a route starts. "me" is this computer (the browser's location); "station" is a saved start; "pick" is a click on the map. */
export type OriginMode = 'me' | 'station' | 'pick';

export interface OriginState {
  mode: OriginMode;
  picked: LatLon | null;
  station: LatLon | null;
  /** The next map click sets `picked`. */
  picking: boolean;
}

export const NO_ORIGIN: OriginState = { mode: 'me', picked: null, station: null, picking: false };

/** The chosen start, or null for "ask the browser where this computer is". */
export const originOf = (o: OriginState): LatLon | null => (o.mode === 'station' ? o.station : o.mode === 'pick' ? o.picked : null);

/** True when the mode needs a point that has not been set yet. */
export const originMissing = (o: OriginState) => o.mode !== 'me' && originOf(o) === null;

export const googleDirectionsUrl = (to: LatLon, from?: LatLon | null) =>
  // No origin given: Google Maps starts from the viewer's current location and lists alternatives.
  `https://www.google.com/maps/dir/?api=1${from ? `&origin=${from.lat},${from.lon}` : ''}&destination=${to.lat},${to.lon}&travelmode=driving`;

export function describeRoute(r: RouteOption): string {
  const km = (r.distanceM / 1000).toFixed(1);
  const min = Math.max(1, Math.round(r.durationS / 60));
  return `${km} km · ${min} min`;
}

export interface RoutingState {
  status: 'idle' | 'locating' | 'loading' | 'done' | 'error';
  error?: string;
  options: RouteOption[];
  active: number;
}

export const NO_ROUTES: RoutingState = { status: 'idle', options: [], active: 0 };

File: android/app/src/main/kotlin/com/macci/kaalerto/route/Routing.kt
package com.macci.kaalerto.route

import com.macci.kaalerto.data.FeatureSummary
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.min

// Route alternatives from OSRM's free public demo server, ranked by how many CURRENT flooded
// features lie along each route. OSRM knows roads, not floods: a road with no report simply
// looks clear, so this ranks by the reports we have and never certifies a route as safe.
// Port of dashboard/src/lib/routing.ts. Demo-only dependency: the public server is
// best-effort and only routes for cars — self-host OSRM (or change OSRM_URL) beyond a demo.
private const val OSRM_URL = "https://router.project-osrm.org/route/v1/driving"

/** A flooded feature within this many metres of the route line counts as "along" it. */
private const val NEAR_M = 75.0

data class LatLon(val lat: Double, val lon: Double)

data class FloodPoint(val lat: Double, val lon: Double, val severity: String)

data class RouteOption(
    /** (lon, lat) pairs, in travel order. */
    val coords: List<Pair<Double, Double>>,
    val distanceM: Double,
    val durationS: Double,
    val s3: Int,
    val s2: Int,
    val sx: Int,
    val safest: Boolean,
) {
    val floodCount get() = s3 + s2 + sx
}

/** What [fetchRoutes] can come back with; a failure is a value so the UI never sees an exception. */
sealed interface RoutesResult {
    data class Ok(val options: List<RouteOption>) : RoutesResult
    /** No connection, or the routing service was unreachable. */
    data object NoConnection : RoutesResult
    data object NoRoute : RoutesResult
    data class ServiceError(val code: Int) : RoutesResult
}

@Serializable
private data class OsrmResponse(val code: String = "", val routes: List<OsrmRoute> = emptyList())

@Serializable
private data class OsrmRoute(val distance: Double, val duration: Double, val geometry: OsrmGeometry)

@Serializable
private data class OsrmGeometry(val coordinates: List<List<Double>>)

/** A raw OSRM route before ranking. */
internal data class RawRoute(val coords: List<Pair<Double, Double>>, val distanceM: Double, val durationS: Double)

private val json = Json { ignoreUnknownKeys = true }

/** Null when the body is not a usable OSRM answer. */
internal fun parseOsrm(body: String): List<RawRoute>? {
    val parsed = runCatching { json.decodeFromString<OsrmResponse>(body) }.getOrNull() ?: return null
    if (parsed.code != "Ok") return null
    return parsed.routes.map { r ->
        RawRoute(r.geometry.coordinates.map { it[0] to it[1] }, r.distance, r.duration)
    }
}

/** Current flooded spots worth avoiding: S3 (impassable), S2 (impassable for cars), SX (conflicting). */
fun floodPoints(summaries: List<FeatureSummary>): List<FloodPoint> =
    summaries
        .filter { !it.isStale && it.severity in AVOID }
        .map { FloodPoint(it.lat, it.lon, it.severity) }

private val AVOID = setOf("S2", "S3", "SX")

/** Metres from p to segment a-b on a local flat approximation (fine at city scale). */
private fun distToSegmentM(p: LatLon, a: Pair<Double, Double>, b: Pair<Double, Double>): Double {
    val kx = 111_320.0 * cos(Math.toRadians(p.lat))
    val ky = 110_540.0
    val ax = (a.first - p.lon) * kx
    val ay = (a.second - p.lat) * ky
    val bx = (b.first - p.lon) * kx
    val by = (b.second - p.lat) * ky
    val dx = bx - ax
    val dy = by - ay
    val len2 = dx * dx + dy * dy
    val t = if (len2 == 0.0) 0.0 else max(0.0, min(1.0, -(ax * dx + ay * dy) / len2))
    return hypot(ax + t * dx, ay + t * dy)
}

private fun nearRoute(p: FloodPoint, coords: List<Pair<Double, Double>>): Boolean {
    val at = LatLon(p.lat, p.lon)
    for (i in 1 until coords.size) {
        if (distToSegmentM(at, coords[i - 1], coords[i]) <= NEAR_M) return true
    }
    return false
}

private fun score(r: RouteOption) = r.s3 * 3 + r.s2 * 2 + r.sx

/**
 * Counts the flooded spots each route passes within [NEAR_M] of and sorts fewest first, then
 * quickest. Spots at the destination are the reason for going, not an obstacle on the way.
 */
internal fun rankRoutes(raw: List<RawRoute>, floods: List<FloodPoint>, to: LatLon): List<RouteOption> {
    val obstacles = floods.filter {
        hypot(
            (it.lat - to.lat) * 110_540.0,
            (it.lon - to.lon) * 111_320.0 * cos(Math.toRadians(to.lat)),
        ) > NEAR_M
    }
    val options = raw.map { r ->
        val hit = obstacles.filter { nearRoute(it, r.coords) }
        RouteOption(
            coords = r.coords,
            distanceM = r.distanceM,
            durationS = r.durationS,
            s3 = hit.count { it.severity == "S3" },
            s2 = hit.count { it.severity == "S2" },
            sx = hit.count { it.severity == "SX" },
            safest = false,
        )
    }.sortedWith(compareBy({ score(it) }, { it.durationS }))
    return options.mapIndexed { i, o -> if (i == 0) o.copy(safest = true) else o }
}

suspend fun fetchRoutes(from: LatLon, to: LatLon, floods: List<FloodPoint>): RoutesResult = withContext(Dispatchers.IO) {
    val url = "$OSRM_URL/${from.lon},${from.lat};${to.lon},${to.lat}?alternatives=3&overview=full&geometries=geojson&steps=false"
    val connection = try {
        URL(url).openConnection() as HttpURLConnection
    } catch (e: IOException) {
        return@withContext RoutesResult.NoConnection
    }
    try {
        // The public server 403s Android's default "Dalvik/..." agent and asks clients to identify themselves.
        connection.setRequestProperty("User-Agent", "KaAlerto/1.0 (community flood map; hackathon demo)")
        connection.connectTimeout = 8_000
        connection.readTimeout = 12_000
        val status = connection.responseCode
        if (status !in 200..299) return@withContext RoutesResult.ServiceError(status)
        val raw = parseOsrm(connection.inputStream.bufferedReader().use { it.readText() })
        if (raw.isNullOrEmpty()) RoutesResult.NoRoute else RoutesResult.Ok(rankRoutes(raw, floods, to))
    } catch (e: IOException) {
        RoutesResult.NoConnection
    } finally {
        connection.disconnect()
    }
}

fun describeRoute(r: RouteOption): String {
    val km = "%.1f".format(r.distanceM / 1000)
    val min = max(1, (r.durationS / 60).toInt())
    return "$km km · $min min"
}
